### PR TITLE
add monitor metrics

### PR DESCRIPTION
**Pull request by runzhliu**
Fully based on [tf-operator](https://github.com/kubeflow/tf-operator/pull/1018), add the metrics monitor by Prometheus in mpi-operator, which can report the Mpijob metrics like creation, deletion, completion and failure.

### CODE REVIEW

File: cmd/mpi-operator.v1alpha2/app/server.go [hunk 1]
```diff
@@ -17,6 +17,8 @@ package app
 import (
 	"context"
 	"fmt"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
```
Comment (carmark) [inline cmd/mpi-operator.v1alpha2/app/server.go:21]: should you please move these two packages to second part?

File: cmd/mpi-operator.v1alpha2/main.go [hunk 2]
```diff
@@ -16,19 +16,38 @@ package main
...
 	"github.com/golang/glog"
 
 	"github.com/kubeflow/mpi-operator/cmd/mpi-operator.v1alpha2/app"
 	"github.com/kubeflow/mpi-operator/cmd/mpi-operator.v1alpha2/app/options"
 )
 
+func startMonitoring(monitoringPort int) {
+	if monitoringPort != 0 {
+		go func() {
+			glog.Infof("Setting up client for monitoring on port: %s", strconv.Itoa(monitoringPort))
```
Comment (carmark) [inline cmd/mpi-operator.v1alpha2/main.go:33]:
```
glog.Infof("Setting up client for monitoring on port: %d", monitoringPort)
```

what about the above one?

File: cmd/mpi-operator.v1alpha2/main.go [hunk 3]
```diff
@@ -16,19 +16,38 @@ package main
...
+			http.Handle("/metrics", promhttp.Handler())
+			err := http.ListenAndServe(fmt.Sprintf(":%s", strconv.Itoa(monitoringPort)), nil)
```
Comment (carmark) [inline cmd/mpi-operator.v1alpha2/main.go:35]: same as above one.

**Comment by carmark** [comment]
If you update the vendor code, could you please separate this commit into two parts?

File: cmd/mpi-operator.v1alpha2/app/options/options.go [hunk 4]
```diff
@@ -60,6 +61,9 @@ func (s *ServerOption) AddFlags(fs *flag.FlagSet) {
 
 	fs.BoolVar(&s.PrintVersion, "version", false, "Show version and quit")
 
+	fs.IntVar(&s.MonitoringPort, "monitoring-port", 8443,
```
Comment (terrytangyuan) [inline cmd/mpi-operator.v1alpha2/app/options/options.go:64]: Should this be disabled by default in case prometheus is not available?

Comment (carmark) [reply cmd/mpi-operator.v1alpha2/app/options/options.go:64]: I think that's all right, prometheus will fetch data from this port. But the default port of monitoring should be 8080?

Comment (terrytangyuan) [reply cmd/mpi-operator.v1alpha2/app/options/options.go:64]: This implementation returns `glog.Error` if the port is not set up for Prometheus: https://github.com/kubeflow/mpi-operator/pull/215/files#diff-58e97bee9d8731245689342bcce48a7dR37

File: cmd/mpi-operator.v1alpha2/app/server.go (hunk 1 above)
Comment (runzhliu) [reply cmd/mpi-operator.v1alpha2/app/server.go]: Sure ;)

**Comment by runzhliu** [comment]
> If you update the vendor code, could you please separate this commit into two parts?

all right.

**Comment by terrytangyuan** [comment]
Actually I am removing the vendor directories in https://github.com/kubeflow/mpi-operator/pull/218. Please revert the changes related to vendor and then rebase. @runzhliu @carmark

**Comment by runzhliu** [comment]
> Actually I am removing the vendor directories in #218. Please revert the changes related to vendor and then rebase. @runzhliu @carmark

Good move!

File: cmd/mpi-operator.v1alpha2/app/options/options.go (hunk 4 above)
Comment (terrytangyuan) [reply cmd/mpi-operator.v1alpha2/app/options/options.go:64]: @runzhliu See my comment above on `glog.Error`

Comment (runzhliu) [reply cmd/mpi-operator.v1alpha2/app/options/options.go]: @terrytangyuan I have set the default value 0, so the metrics monitor wouldn't be triggered on.

**Pull request merged**